Site Address Changer: Offer Site Redirect Upgrade

Aurorum opened the pull request:
#### Changes proposed in this Pull Request

* Offer the Site Redirect upgrade when a user is changing their site

#### Testing instructions

Each of the options should (hopefully!) be clear. 

1. Should add the Site Redirect upgrade, change the site address, and take you immediately to the Checkout, without deleting your site - it's worth making sure the price is correct too!
2. Should change the site address without deleting the site
3. Should change the site address and delete the site (current default)

| Before                                                                                                                                                                | After                                                                                                                                                                 |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <img width="519" alt="Screenshot 2021-09-02 at 20 08 06" src="https://user-images.githubusercontent.com/43215253/131902305-87b4d0dc-af52-4a3c-913e-9e3deb853934.png"> | <img width="644" alt="Screenshot 2021-09-02 at 19 51 35" src="https://user-images.githubusercontent.com/43215253/131902309-b52a00a5-1e60-432f-ae31-4799fcd0f264.png"> |

@supernovia, @formosattic - let me know if you have any thoughts! 

Fixes #55801

cc @jsnajdr

Aurorum committed:
Add Site Redirect to cart

Aurorum committed:
Update dialog

Aurorum committed:
Include styling

Aurorum committed:
Fix redirect

supernovia commented:
I can't test it, but it looks great! Let me see if we can get more eyes on this.

Aurorum commented:
Thank you!

simison commented:
Are all those 3 options truly non-reversable like the checkbox indicates?

simison commented on client/blocks/site-address-changer/dialog.jsx:
```diff
@@ -98,7 +146,7 @@ class SiteAddressChangerConfirmationDialog extends PureComponent {
 					<p className="site-address-changer__confirmation-detail-copy site-address-changer__copy-deletion">
 						{ translate(
-							'{{strong}}%(currentDomainName)s{{/strong}}%(currentDomainSuffix)s will be removed and unavailable for use.',
+							'{{strong}}%(currentDomainName)s{{/strong}}%(currentDomainSuffix)s will no longer be your site address.',
```
Worth applying "string freeze" label in the PR to ensure we have translations before deploying.

Aurorum replied:
I don't have the ability to add labels - would you mind please doing this? :)

simison replied:
Someone reviewing the work can apply once we know they're final. :-)

Aurorum commented:
I think it might be possible to contact support at present to get the change reversed, but other than that, I think the the actual site address change is irreversible - the options might not be, since you can always cancel a Site Redirect upgrade. However, I think that checkbox conveys that as soon as the confirm button is pressed, the site address change will take place and there's no simple "undo" better.

supernovia commented:
If someone changes their site URL, all of the content, followers, stats, customization, etc all go to the new address. The old address, currently, just gets deleted.

Support could un-delete the site, but it would be empty. To undo the changes, someone would need to move everything back to the old address. So in that sense it really is undoable unless we're willing to do a lot of manual work.

rafaelgallani commented on client/blocks/site-address-changer/dialog.jsx:
```diff
@@ -68,11 +81,46 @@ class SiteAddressChangerConfirmationDialog extends PureComponent {
+		const options = [
+			{
+				id: 'redirect',
+				answerText: translate(
+					'Redirect to {{strong}}%(newSiteAddress)s{{/strong}} (%(price)s per year)',
+					{
+						components: {
+							strong: <strong />,
+						},
+						args: {
+							newSiteAddress: newDomainName + newDomainSuffix,
+							price: siteRedirectPrice,
+						},
+					}
+				),
```
When loading the dialog, the prop type checker yields an error:

![change-site-address-dev-console-error](https://user-images.githubusercontent.com/18705930/132250965-494b2e5d-2c29-4a23-b2a3-5ba98d02e12a.png)

Here's how the component props are in that moment:

![change-site-address-dev-console-react-component-error](https://user-images.githubusercontent.com/18705930/132251033-e474041f-2760-4f45-b45e-4c42dbec6a5c.png)

This happens because `translate` returns an array in this case (a React Fragment). 

While this is not the optimal solution, it can be easily solved by removing the `components` prop, thus returning a string. 
Still, I think the best way to fix this would be to change the prop type on the `MultipleChoiceQuestion` component.
Leaving it as-is will only show the error on the console. The component is rendered just fine 😸

rafaelgallani commented on client/blocks/site-address-changer/index.jsx:
```diff
@@ -71,8 +77,23 @@ export class SiteAddressChanger extends Component {
+			this.props.shoppingCartManager
+				.addProductsToCart( [
+					fillInSingleCartItemAttributes(
+						siteRedirect( { domainFieldValue } ),
```
I was not able to complete the checkout, an error was returned from the API:

![image](https://user-images.githubusercontent.com/18705930/132251330-9073d15e-3232-4051-a910-67160943d87d.png)

Perhaps because the properties here are wrong? I don't see `domainFieldValue` as a property of `siteRedirect`. I think it should be `domain`: 
```suggestion
						siteRedirect( { domain: domainFieldValue } ),
```

rafaelgallani reviewed (changes requested):
Hey @Aurorum! Thanks for the PR!
I tested it locally and left a few comments on some points 😄 

Also, a more general question: 
When selecting the site redirect option, should we change the site address **before** checking out???
I think it would make more sense to add it **after** the checkout. 
Otherwise, if an error or anything else occurs (_like it did when I was testing locally_), the site would be changed already, making the user confused.

Aurorum commented:
Thanks for the review @rafaelgalani - I'll fix those issues soon! 

> When selecting the site redirect option, should we change the site address before checking out???

Yes, I think that's reasonable. Would you mind providing me with a screenshot of the screen which appears after the Site Redirect checkout is complete? I can't seem to find the function for when a purchase is made in the code, and I only have a limited number of credits to play with, so I think it'd be better to try coming up with a solution before testing it first. Thank you!

supernovia commented:
> I only have a limited number of credits to play with, so I think it'd be better to try coming up with a solution before testing it first. Thank you!

I don't have a test environment but I can help with this part!

Aurorum commented:
That'd be great if you could send a screenshot - I should be able to figure it out from there!

supernovia commented:
> That'd be great if you could send a screenshot - I should be able to figure it out from there!

Oh! Clarifying, don't have a test environment, but I did put in some credit so you'll have more to work with.

Aurorum committed:
Merge branch 'trunk' into update/site-address-changer

Aurorum commented:
Ah, oops - thank you!

Aurorum commented:
Just an update: I had this completed, but my solution relied on changing the site address immediately after completing the purchase. Unfortunately, it turns out that you won't be able to purchase the Site Redirect upgrade before your site address has been changed, so my solution doesn't work. This causes complications when trying to implement changing the site address after checking out. 

I think that the site address needs to be changed *before* the transaction occurs, but *after* the user has clicked the "Pay now" button, so probably in this file: https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/lib/submit-wpcom-transaction.ts

The issue is I'm not that familiar with Typescript, so I'm not sure how to make the `requestSiteAddressChange` function work in there. I imagine there's probably something very simple which I'm missing, so if anyone would know how to make that request work, please let me know!

jsnajdr commented on client/blocks/site-address-changer/index.jsx:
```diff
@@ -18,13 +21,17 @@ import {
 	clearValidationError,
 } from 'calypso/state/site-address-change/actions';
+import { fillInSingleCartItemAttributes } from 'calypso/lib/cart-values';
+import { hasProduct, siteRedirect } from 'calypso/lib/cart-values/cart-items';
```
The `hasProduct` import seems to be not used. There are more ESLint errors like this being reported on both the `index.jsx` and `dialog.jsx` files.

jsnajdr commented on client/blocks/site-address-changer/dialog.jsx:
```diff
@@ -2,9 +2,13 @@ import { Dialog, Gridicon } from '@automattic/components';
 import FormInputCheckbox from 'calypso/components/forms/form-checkbox';
 import FormLabel from 'calypso/components/forms/form-label';
 import TrackComponentView from 'calypso/lib/analytics/track-component-view';
+import MultipleChoiceQuestion from 'calypso/components/multiple-choice-question';
```
The `MultipleChoiceQuestion` component is more fit for some marketing surveys rather than site settings. I would personally build this part of the form from `FormFieldset` and `FormRadio` components.

jsnajdr commented on client/blocks/site-address-changer/dialog.jsx:
```diff
@@ -67,11 +80,46 @@ class SiteAddressChangerConfirmationDialog extends PureComponent {
+			},
+			{
+				id: 'no-redirect',
+				answerText: translate( 'Keep address without redirecting it' ),
```
It's unclear to me what exactly does "keep address" mean. It doesn't redirect to the new name, it's not deactivated... Which site does it point to then?

jsnajdr reviewed (commented):
The PR should be rebased before further review -- there's quite a few conflicts, we newly enforce import ordering, there are a few more ESLint warnings... It's not in a very clean state.

Aurorum commented:
Just noting that I will revisit this again over the weekend - thanks for taking a look!

Aurorum committed:
Merge branch 'trunk' into update/site-address-changer

Aurorum committed:
Address initial review

Aurorum commented:
I've addressed the review above, namely through building the `MultipleChoiceQuestion` component and adding some explanations, but still unsure about how to deal with the issue mentioned in https://github.com/Automattic/wp-calypso/pull/55966#issuecomment-917362248.

<img width="674" alt="Screenshot 2021-11-07 at 16 58 59" src="https://user-images.githubusercontent.com/43215253/140654333-3d3dc25e-6a6e-449f-8725-2192d9684e14.png">

jsnajdr commented on client/state/site-address-change/actions.js:
```diff
@@ -151,11 +152,13 @@ export const requestSiteAddressChange = (
-			// new name of the `*.wordpress.com` domain that we just changed
-			const newDomain = newSlug + '.' + domain;
-			page( domainManagementEdit( siteSlug, newDomain ) );
+			if ( shouldRedirect ) {
```
The name of this param seems to be wrong: the value passed is `option !== 'redirect'`, i.e., it means "should _not_ redirect".

An even better solution would be to move the "redirect to domain edit" logic out of the action, and chain promises at the call site instead:
```js
this.props.requestSiteAddressChange().then( () => {
  if ( options !== 'redirect' ) {
    page( domainManagementEdit() );
  }
} );
```

jsnajdr commented on client/blocks/site-address-changer/index.jsx:
```diff
@@ -70,8 +76,23 @@ export class SiteAddressChanger extends Component {
-			type
+			type,
+			option === 'discard',
+			option !== 'redirect'
```
If we want to change the site address before adding the redirect to the cart and going to checkout (I'm not sure if that's the right order!), there should be some `await` or `.then` here to wait for the address change REST request to finish before proceeding.

jsnajdr reviewed (commented):
> Unfortunately, it turns out that you won't be able to purchase the Site Redirect upgrade before your site address has been changed, so my solution doesn't work.

To clarify: at this moment, the workflow is that you:
1. choose the "redirect" radio box
2. click "Change site address" button
3. the REST request to change address is sent, while still on the "site address change" page
4. site redirect is added to the cart
5. you're navigated from the "site address change" page to the "checkout" page
6. you finish the purchase by clicking "Pay now"
7. REST request to `/me/transactions` is sent
8. and after checkout you are redirected somewhere, I don't know where?

And you would like to do the REST request with site address change at some later time? At step 7, immediately before the `/me/transactions` request?

And the reason is to do the irreversible site address change as late as possible, when we know that the user is really about to pay and do the transaction?

I'm personally not very familiar with the checkout and payment workflows. Cc @sirbrillig for awareness. After we figure out what workflow we really want, we'll need some advice with implementing it at the right place in the `composite-checkout` flow.

Aurorum commented:
> And you would like to do the REST request with site address change at some later time? At step 7, immediately before the /me/transactions request? And the reason is to do the irreversible site address change as late as possible, when we know that the user is really about to pay and do the transaction?

Yes, that's both correct.

sirbrillig commented:
> and after checkout you are redirected somewhere, I don't know where?

It depends on [a great many factors](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts). 

> After we figure out what workflow we really want, we'll need some advice with implementing it at the right place in the composite-checkout flow. ... At step 7, immediately before the /me/transactions request

I don't fully grok what's happening here, but we do something similar to what you describe with the account creation request for [stripe](https://github.com/Automattic/wp-calypso/blob/41704afe7ae693c8b1ba3d0321d95292a21be2d7/client/my-sites/checkout/composite-checkout/lib/submit-wpcom-transaction.ts#L18) and [paypal](https://github.com/Automattic/wp-calypso/blob/41704afe7ae693c8b1ba3d0321d95292a21be2d7/client/my-sites/checkout/composite-checkout/lib/paypal-express-processor.ts#L79) transactions.

jsnajdr commented:
> we do something similar to what you describe with the account creation request 

Yes, that looks like a great blueprint for the site address change, too. Do the irreversible/expensive action as late as possible, just before submitting to the `/me/transaction` endpoint that actually pays for the performed action.

lupus2k commented:
Good job @Aurorum, but since this PR is more than 1 year old and there have been several changes in the code, we need you to rebase against trunk, can you do that?
